python_bindings/test/correctness/pystub.py

```diff
@@ -270,9 +264,12 @@ def test_complex(cls, extra_input_name = ""):
 
 
 if __name__ == "__main__":
-    test_simple(simple_pystub)
-    test_complex(complex_pystub)
-    test_complex(complex_pystub, extra_input_name = "foozz_input")
-    test_simple(SimplePy)
-    test_complex(ComplexPy)
-    test_complex(ComplexPy, extra_input_name = "foo_input")
+    target = hl.get_jit_target_from_environment()
+    ctx = hl.GeneratorContext(target)
+    with hl.set_context(ctx):
```

Comment:
Yeah, this is a bit awkward... is it not possible to implement this stuff on `GeneratorContext` such that it adjusts the context and resets it later? Something like this might work:

```python
class GeneratorContext:
    # ... stuff ...
    def __enter__(self):
        token = _halide_generator_context.set(self)
        class _ResetToken:
            def __exit__(self):
                _halide_generator_context.reset(token)
        return _ResetToken()
```
Then you could write:
```python
with hl.GeneratorContext(target):
    # ...
```

Reply:
An alternative approach could be to write:
```python
@contextmanager
def set_context(*args, **kwargs):
    context = hl.GeneratorContext(*args, **kwargs)
    token = _halide_generator_context.set(context)
    try:
        yield context
    finally:
        _halide_generator_context.reset(token)
```

and then the idiom would be:
```python
with hl.set_context(target) as context:
    # ...
```

Reply:
Ack, no, my first suggestion doesn't work. Python calls __exit__ on the original object, not the one that is returned by __enter__.

Reply:
> is it not possible to implement this stuff on GeneratorContext

I tried a lot of things yesterday and couldn't see an obvious way to make it work, but yes, I would greatly prefer that syntax

> Something like this might work

Not sure I understand how that could work: for one, we're returning an instance of _ResetToken (so the `with ctx as c` variant would have c as a _ResetToken rather than a GeneratorContext). But more problematically, I don't see how __exit__ would ever get called; if I read PEP 343 correctly, the __exit__ method is always called on the original object in the with statement, not on the value returned.

> An alternative approach could be to write:

I'm not sure I understand how this is different from what I already have here? (aside from the fact that it implicitly converts a Target into a GeneratorContext, which would be handy to do but trivial to slot into what is already here)

Reply:
> You could still do this, though (tested):

Yeah... I guess that would work, though I must admit it makes me feel kinda squicky for reasons I can't put my finger on. I'll give it a try.

Reply:
> makes me feel kinda squicky for reasons I can't put my finger on

Me too... I'm trying to find a case where "normal" code calls `__enter__` and `__exit__` out of sync, but this might just be paranoid. I feel like API ergonomics should trump sharp corners that are only exposed if you peel off the warranty sticker.

Reply:
Revised, PTAL